src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst

```diff
@@ -0,0 +1,557 @@
+    let ((), tr') = trigger_event sender (CommConfSendMsg sender receiver payload_bytes) tr' in
+    encrypt_message_proof tr' sender receiver pk_receiver nonce payload;
+    let msg_encrypted = encrypt_message pk_receiver nonce payload in
+    let (msg_id, tr') = send_msg msg_encrypted tr' in
+    assert(tr_out == tr');
+    ()
+  )
+
+val decrypt_message_proof:
+  {|cinvs:crypto_invariants|} ->
+  #a:Type -> {| parseable_serializeable bytes a |} ->
+  tr:trace ->
+  receiver:principal -> sk_receiver:bytes -> msg_encrypted:bytes ->  
+  Lemma
+  (requires
+    has_communication_layer_crypto_predicates /\
+    is_private_key_for tr sk_receiver (LongTermPkeKey comm_layer_pkenc_tag) receiver /\
+    bytes_invariant tr msg_encrypted
+  )
+  (ensures
+    (
+      match decrypt_message #a sk_receiver msg_encrypted with
+      | None -> True
+      | Some payload -> (exists sender.
+        is_well_formed a (is_knowable_by (comm_label sender receiver) tr) payload /\
+        is_knowable_by (comm_label sender receiver) tr (serialize a payload) /\
+        (
+          event_triggered tr sender (CommConfSendMsg sender receiver (serialize a payload)) \/
+          is_well_formed a (is_publishable tr) payload
+        )
+      )
+    )    
+  )
+let decrypt_message_proof #cinvs #a tr receiver sk_receiver msg_encrypted =
+  match decrypt_message #a sk_receiver msg_encrypted with
+  | None -> ()
+  | Some payload -> (
+    let Some plaintext = pke_dec sk_receiver msg_encrypted in
+    serialize_parse_inv_lemma #bytes a plaintext;
+    ()
+  )
+
+val receive_confidential_proof:
+  {|invs:protocol_invariants|} ->
+  #a:Type -> {| parseable_serializeable bytes a |} ->
+  tr:trace ->
+  higher_layer_preds:comm_higher_layer_event_preds a ->
+  comm_keys_ids:communication_keys_sess_ids ->
+  receiver:principal -> msg_id:timestamp ->
+  Lemma
+  (requires
+    trace_invariant tr /\ 
+    has_private_keys_invariant /\
+    has_communication_layer_crypto_predicates /\
+    has_communication_layer_event_predicates higher_layer_preds
+  )
+  (ensures (
+    match receive_confidential #a comm_keys_ids receiver msg_id tr with
+    | (None, tr_out) -> trace_invariant tr_out
+    | (Some payload, tr_out) -> ( 
+      trace_invariant tr_out /\
+      (
+        (exists sender.
+          is_well_formed a (is_knowable_by (comm_label sender receiver) tr) payload /\
+          // We explicitly do not use `event_triggered tr receiver
+          // (CommConfReceiveMsg receiver (serialize a payload))` as a
+          // post-condition here. The reason for the following form of the
+          // post-condition is that it allows us on the higher layer to assert a
+          // statement `b` contained in the `higher_layer_preds` that depends on
+          // some field `field1` of the payload in the following way: `b
+          // payload.field1 \/ is_publishable tr payload.field1`.
+          event_triggered tr sender (CommConfSendMsg sender receiver (serialize a payload))
```

Comment:
It's not entirely obvious to me why this is needed --- my intuition is that `CommConfReceiveMsg` existing implies a corresponding earlier `CommConfSendMsg` must exist (or that the payload is publishable), and so this would seem to me to be a slightly stronger guarantee with `ReceiveMsg` instead. Does that implication not actually hold, or is there some other technical problem?

Reply:
I think this is a technical issue but I did not further investigate it because this form worked for me.

Reply:
Resolved by #101